pyface/data_view/data_models/array_data_model.py

```diff
@@ -0,0 +1,298 @@
+# (C) Copyright 2005-2020 Enthought, Inc., Austin, TX
+# All rights reserved.
+#
+# This software is provided without warranty under the terms of the BSD
+# license included in LICENSE.txt and may be redistributed only under
+# the conditions described in the aforementioned license. The license
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+# Thanks for using Enthought open source!
+""" Provides an N-dimensional array data model implementation.
+
+This module provides a concrete implementation of a data model for an
+n-dim numpy array.
+"""
+from collections.abc import Sequence
+
+from traits.api import Array, HasRequiredTraits, Instance, observe
+
+from pyface.data_view.abstract_data_model import AbstractDataModel
+from pyface.data_view.abstract_value_type import AbstractValueType
+from pyface.data_view.value_types.api import (
+    ConstantValue, FloatValue, IntValue, TextValue, no_value
+)
+from pyface.data_view.index_manager import TupleIndexManager
+
+
+class _AtLeastTwoDArray(Array):
+    """ Trait type that holds an array that at least two dimensional.
+
+    This calls numpy.atleast_2d during validation to ensure that the value is
+    good.
+    """
+
+    def validate(self, object, name, value):
+        from numpy import atleast_2d
+        value = super().validate(object, name, value)
+        return atleast_2d(value)
```

Comment:
I'm happy with it not raising an exception right now: the case that I was most concerned about was actually if no data is provided (we get an empty view); and a single column is a sane default for a 1D array (way better than a single row if the array is large).  If someone really doesn't like it, they can provide data in the shape that they really want without too much difficulty.

I think that these are the right default choices.

Reply:
> a single column is a sane default for a 1D array (way better than a single row if the array is large). 

Wait, given an array of shape `(X, Y)`, isn't X the number of rows and Y the number of columns?
If the first dimension is the number of rows, I think `atleast_2d` returns a single row, rather than a single column.

> If someone really doesn't like it, they can provide data in the shape that they really want without too much difficulty.

The same can be applied to when a validation occurs, it is easy to put it right. Trying to guess an orientation when a 1D array is provided is "guessing-what-you-mean", which is not a good thing to do.

Reply:
Argh. You're right, it does produce a single row.  That's not optimal.  I'll fix it one way or another.

In terms of APIs, I feel that there are different standards depending on who the user of the API is.  In this case `ArrayDataModel` is designed to be used by other programmers who want to display their data, so there is value in affordances where 90%+ of the time "what you mean" is clear, particularly if the other 10% is easy to make the meaning clear (ie. we aren't _preventing_ the other 10% of uses).  By saving some effort for the 90% case, you make the API more ergonomic.

There is an issue which this has raised though, which is that `atleast_2d` doesn't modify in place (good!) but it's not the original array that is assigned (bad!) - I have to have another look at what implicit promises `Array` makes in terms of identity.

Reply:
Following up on the last point - for better or worse, the validate method of the `Array` trait does not guarantee that the object stored in the trait is the same object that was set, even in the non-coercing case.  In particular there are places where `asarray` is called, which may create new `ndarray` instances.

Reply:
And following up on the last point; `traits.trait_numeric` _really_ needs a clean-up or re-implementation.

Reply:
Woops, I did not notice that the array may be a copy. Agreed that's bad, I think we want to keep the original array.
This is quite similar to the problems / surprises people often run into when they assign a list to a trait of `List` type... the list is copied implicitly.

Maybe we just want an instance of array here for now. Developers should make sure they assign an array with at least two dimensions: If they don't, bad things will happen anyway, it just happens later (less ideal). We can introduce the validation logic later after cleaning up `trait_numeric`. It seems more important to be making sure the original array is used.

Reply:
I think `trait_numeric` is going to need a re-write.  The name is a hint: Numeric was the predecessor to NumPy.

But even if we use `Array` we aren't giving any guarantees for the user on identity of the array.  I've simplified things out, but I'm also happy to revert to `Array` with a shape (0, 0) default supplied by a default handler with the idea of fixing later.